### PR TITLE
Always run coverage against base

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR modifies our github actions to always run coverage against the base branch.

**This workflow assumes the existence of a cached `solidity/coverage` directory keyed by the sha of the base branch. If no cache entry is found, the workflow will fail**

Because there won't be a cache entry for main, we hacked around this by temporarily committing the current coverage artifacts, which causes the caching action to cache those as if they were the values on main. They were then removed.

This workflow is also updated to fail iff both hardhat and forge coverage decrease.

### Drive-by changes

- None
### Related issues

- None

### Backward compatibility

Yes

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

None

### Testing

Manual - running the action

### CODE REVIEW

**Comment by asaj** [comment]
Open to requiring that forge coverage never decrease, wdyt?

File: .github/workflows/node.yml [hunk 1]
```diff
@@ -21,11 +21,15 @@ jobs:
...
         run: git submodule foreach rm yarn.lock
 
-      - uses: actions/cache@v3
+      - name: yarn-cache
+        uses: actions/cache@v3
         with:
           path: '**/node_modules'
           key: ${{ runner.os }}-yarn-cache4-${{ hashFiles('./yarn.lock') }}
 
+      - name: Dump GitHub base sha
```
Comment (yorhodes) [inline .github/workflows/node.yml:30]: remove?

File: solidity/.gitignore [hunk 2]
```diff
@@ -6,6 +6,7 @@ dist/
 coverage/
 coverage.json
 .env
+*lcov.base.info
```
Comment (yorhodes) [inline solidity/.gitignore:9]: should this include *lcov.info also

File: .github/workflows/node.yml [hunk 3]
```diff
@@ -132,19 +136,46 @@ jobs:
...
+
+      - name: check-coverage
+        run: |
+          sudo apt-get install lcov
+          cd ./solidity
+          export H_BASE=$(lcov --summary ./hardhat-lcov.base.info | grep "lines" | sed 's/.*lines......:\ //' | sed 's/% (.*//')
+          export H_REF=$(lcov --summary ./coverage/lcov.info | grep "lines" | sed 's/.*lines......:\ //' | sed 's/% (.*//')
+          export F_BASE=$(lcov --summary ./forge-lcov.base.info | grep "lines" | sed 's/lines......://' | sed 's/% (.*//')
+          export F_REF=$(lcov --summary ./coverage/forge-lcov.info | grep "lines" | sed 's/lines......://' | sed 's/% (.*//')
+          if [[ $H_REF < $H_BASE ]] && [[ $F_REF < $F_BASE ]]; then exit 1; fi
```
Comment (yorhodes) [inline .github/workflows/node.yml:181]: surprised the action doesn't support failing on diff > x%

**Review by yorhodes — APPROVED** [review]

File: .github/workflows/node.yml (hunk 1 above)
Comment (asaj) [reply .github/workflows/node.yml:30]: Oops

File: solidity/.gitignore (hunk 2 above)
Comment (asaj) [reply solidity/.gitignore:9]: Doesn't matter, I don't think we write *lcov.info to ./soldiity

File: .github/workflows/node.yml (hunk 3 above)
Comment (asaj) [reply .github/workflows/node.yml:181]: It supports a minimum coverage, I suppose the best thing would be to plumb the base coverage into the minimum coverage, but I have it working this way so will keep it as is for now

Comment (yorhodes) [reply .github/workflows/node.yml:178]: https://github.com/romeovs/lcov-reporter-action/issues/35

Comment (asaj) [reply .github/workflows/node.yml:178]: @yorhodes what you linked above seems worse than caching? Going to merge now so I don't have to re-check in new artifacts but happy to follow up

**Pull request merged**